utils/entities/nft/erc-1155.ts

```diff
@@ -0,0 +1,65 @@
+import { NftType } from '../../enums'
+import { Account, Nft } from '../../../generated/schema'
+import { findOrCreateAccount } from '../account'
+import { Address, BigInt, store } from '@graphprotocol/graph-ts'
+
+export function generateERC1155NftId(tokenAddress: string, tokenId: BigInt, ownerAddress: string): string {
+  return tokenAddress + '-' + tokenId.toString() + '-' + ownerAddress
+}
+
+export function upsertERC1155Nft(tokenAddress: string, tokenId: BigInt, amount: BigInt, from: string, to: string): Nft {
+  const nftId = generateERC1155NftId(tokenAddress, tokenId, to)
+
+  let nft = Nft.load(nftId)
+  if (!nft) {
+    nft = new Nft(nftId)
+    nft.tokenId = tokenId
+    nft.tokenAddress = tokenAddress
+    nft.owner = findOrCreateAccount(to).id
+    nft.type = NftType.ERC1155.toString()
+  }
+
+  updateERC1155Balance(findOrCreateAccount(from), findOrCreateAccount(to), nft, amount)
+
+  return nft
+}
+
+export function findOrCreateERC1155Nft(tokenAddress: string, tokenId: BigInt, to: Account): Nft {
+  const nftId = generateERC1155NftId(tokenAddress, tokenId, to.id)
+
+  let nft = Nft.load(nftId)
+  if (!nft) {
+    nft = new Nft(nftId)
+    nft.tokenId = tokenId
+    nft.tokenAddress = tokenAddress
+    nft.type = NftType.ERC1155.toString()
+    nft.owner = to.id
+    nft.save()
+  }
+
+  return nft
+}
+
+export function updateERC1155Balance(from: Account, to: Account, toNft: Nft, amount: BigInt): void {
+  if (from.id == to.id) return // should never happen
+
+  if (from.id != Address.zero().toHex()) {
+    const fromNft = findOrCreateERC1155Nft(toNft.tokenAddress, toNft.tokenId, from)
+    if (!fromNft.amount) return store.remove('Nft', fromNft.id) // should never happen
+
+    // remove if the decudecting amount is 0
+    if (fromNft.amount!.minus(amount).equals(BigInt.fromI32(0))) {
+      store.remove('Nft', fromNft.id)
+    } else {
+      fromNft.amount = fromNft.amount!.minus(amount)
```

Comment:
Do we need null assertion here? We already checked that is not null in a previous `if`
```suggestion
      fromNft.amount = fromNft.amount.minus(amount)
```

Reply:
subgraph compiler is not smart enough to realize that amount is not null here, that's why we need to put a `!` here. When i removed it, this error is thrown:
<img width="1073" alt="Screenshot 2023-10-27 at 10 07 17" src="https://github.com/OriumNetwork/roles-subgraph/assets/38267570/64204c28-f3f1-4848-b420-0237c285c27d">